File: mobile/lib/services/auth_service.dart
import '../models/user_model.dart';
import 'api_service.dart';
import 'storage_service.dart';

class AuthService {
  final _api = ApiService();
  final _storage = StorageService();

  Future<Map<String, dynamic>> register(Map<String, dynamic> userData) async {
    try {
      final response = await _api.post('/auth/register', userData);
      
      if (response['success'] == true) {
        final user = UserModel.fromJson(response['data']['user']);
        final token = response['data']['token'];
        
        await _storage.saveToken(token);
        await _storage.saveUser(user);
        
        return {'success': true, 'user': user};
      }
      
      throw Exception(response['message'] ?? 'Registration failed');
    } catch (e) {
      return {'success': false, 'error': e.toString()};
    }
  }

  Future<Map<String, dynamic>> login(String email, String password) async {
    try {
      final response = await _api.post('/auth/login', {
        'email': email,
        'password': password,
      });
      
      if (response['success'] == true) {
        final user = UserModel.fromJson(response['data']['user']);
        final token = response['data']['token'];
        
        await _storage.saveToken(token);
        await _storage.saveUser(user);
        
        return {'success': true, 'user': user};
      }
      
      throw Exception(response['message'] ?? 'Login failed');
    } catch (e) {
      return {'success': false, 'error': e.toString()};
    }
  }

  Future<void> logout() async {
    await _storage.clearAll();
  }

  Future<UserModel?> getCurrentUser() async {
    try {
      final token = await _storage.getToken();
      if (token == null) return null;

      final response = await _api.get('/auth/me');
      
      if (response['success'] == true) {
        final user = UserModel.fromJson(response['data']['user']);
        await _storage.saveUser(user);
        return user;
      }
      
      return null;
    } catch (e) {
      return null;
    }
  }

  UserModel? getCachedUser() {
    return _storage.getUser();
  }

  // Verify email
  Future<Map<String, dynamic>> verifyEmail(String token) async {
    try {
      final response = await _api.get('/auth/verify-email?token=$token');
      return {
        'success': response['success'] == true,
        'message': response['message'] ?? 'Email verified successfully',
      };
    } catch (e) {
      return {'success': false, 'error': e.toString()};
    }
  }

  // Resend verification email
  Future<Map<String, dynamic>> resendVerification(String email) async {
    try {
      final response = await _api.post('/auth/resend-verification', {'email': email});
      return {
        'success': response['success'] == true,
        'message': response['message'] ?? 'Verification email sent',
      };
    } catch (e) {
      return {'success': false, 'error': e.toString()};
    }
  }

  // Forgot password
  Future<Map<String, dynamic>> forgotPassword(String email) async {
    try {
      final response = await _api.post('/auth/forgot-password', {'email': email});
      return {
        'success': response['success'] == true,
        'message': response['message'] ?? 'Password reset email sent',
      };
    } catch (e) {
      return {'success': false, 'error': e.toString()};
    }
  }

  // Reset password
  Future<Map<String, dynamic>> resetPassword(String token, String password) async {
    try {
      final response = await _api.post('/auth/reset-password', {
        'token': token,
        'password': password,
      });
      return {
        'success': response['success'] == true,
        'message': response['message'] ?? 'Password reset successfully',
      };
    } catch (e) {
      return {'success': false, 'error': e.toString()};
    }
  }
}

File: web/src/services/authService.js
import api from './api'
import axios from 'axios'
import { API_URL } from '../utils/constants'

// Create separate axios instance for login with reasonable timeout
const loginApi = axios.create({
  baseURL: API_URL,
  headers: {
    'Content-Type': 'application/json',
  },
  timeout: 20000, // 20 seconds for login (increased to allow for slow DB connections)
})

// Add request interceptor for login API
loginApi.interceptors.request.use(
  (config) => {
    // No token needed for login
    return config
  },
  (error) => {
    return Promise.reject(error)
  }
)

// Add response interceptor for login API
loginApi.interceptors.response.use(
  (response) => response.data,
  (error) => {
    let message = 'Login failed'
    
    if (error.code === 'ECONNABORTED' || error.message.includes('timeout')) {
      message = 'Login request timed out. Please check your connection and try again.'
    } else if (error.response) {
      message = error.response?.data?.message || error.message || 'Login failed'
    } else if (error.request) {
      message = 'No response from server. Please check your connection.'
    } else {
      message = error.message || 'Login failed'
    }
    
    return Promise.reject(new Error(message))
  }
)

export const authService = {
  // Register new user
  register: async (userData) => {
    const response = await api.post('/auth/register', userData)
    // Note: api interceptor already unwraps response.data, so response IS the data
    if (response.token) {
      localStorage.setItem('token', response.token)
      localStorage.setItem('user', JSON.stringify(response.user))
    }
    return response
  },

  // Login user - use faster login API instance
  login: async (credentials) => {
    const response = await loginApi.post('/auth/login', credentials)
    // Note: loginApi interceptor already unwraps response.data, so response IS the data
    if (response.token) {
      localStorage.setItem('token', response.token)
      localStorage.setItem('user', JSON.stringify(response.user))
    }
    return response
  },

  // Logout user
  logout: () => {
    localStorage.removeItem('token')
    localStorage.removeItem('user')
  },

  // Get current user
  getCurrentUser: async () => {
    const response = await api.get('/auth/me')
    // Note: api interceptor already unwraps response.data
    return response.user
  },

  // Verify token
  verifyToken: async (token) => {
    const response = await api.post('/auth/verify-token', { token })
    // Note: api interceptor already unwraps response.data
    return response
  },

  // Get stored token
  getToken: () => {
    return localStorage.getItem('token')
  },

  // Get stored user
  getStoredUser: () => {
    const user = localStorage.getItem('user')
    return user ? JSON.parse(user) : null
  },

  // Verify email
  verifyEmail: async (token) => {
    const response = await api.get('/auth/verify-email', { params: { token } })
    return response
  },

  // Resend verification email
  resendVerification: async (email) => {
    const response = await api.post('/auth/resend-verification', { email })
    return response
  },

  // Forgot password
  forgotPassword: async (email) => {
    const response = await api.post('/auth/forgot-password', { email })
    return response
  },

  // Reset password
  resetPassword: async (token, password) => {
    const response = await api.post('/auth/reset-password', { token, password })
    return response
  },
}

